api.go

```diff
@@ -15,236 +15,285 @@
-		return errorResponse(reason, errorCode)
+		errResponse := errorResponse(reason, errorCode, http.StatusInternalServerError)
+		return nil, &errResponse
 	}
 
+	dataWrapper.verifyApiKeySuccessResponse.ApiProduct = shortListApiProduct(dataWrapper.apiProducts, verifyApiKeyReq)
 	/*
-	 * Perform all validations related to the Query made with the data
-	 * we just retrieved
+	 * Perform all validations
 	 */
-	result := validatePath(resName.String, path)
-	if result == false {
-		reason := "Path Validation Failed (" + resName.String + " vs " + path + ")"
-		errorCode := "PATH_VALIDATION_FAILED"
-		return errorResponse(reason, errorCode)
+	errResponse := apiM.performValidations(dataWrapper)
+	if errResponse != nil {
+		return nil, errResponse
+	}
+
+	apiM.enrichAttributes(&dataWrapper)
+
+	setDevOrCompanyInResponseBasedOnCtype(dataWrapper.ctype, dataWrapper.tempDeveloperDetails, &dataWrapper.verifyApiKeySuccessResponse)
 
+	return &dataWrapper.verifyApiKeySuccessResponse, nil
+}
+
+func setDevOrCompanyInResponseBasedOnCtype(ctype string, tempDeveloperDetails DeveloperDetails, response *VerifyApiKeySuccessResponse) {
+	if ctype == "developer" {
+		response.Developer = tempDeveloperDetails
+	} else {
+		response.Company = CompanyDetails{
+			Id:             tempDeveloperDetails.Id,
+			Name:           tempDeveloperDetails.FirstName,
+			DisplayName:    tempDeveloperDetails.UserName,
+			Status:         tempDeveloperDetails.Status,
+			CreatedAt:      tempDeveloperDetails.CreatedAt,
+			CreatedBy:      tempDeveloperDetails.CreatedBy,
+			LastmodifiedAt: tempDeveloperDetails.LastmodifiedAt,
+			LastmodifiedBy: tempDeveloperDetails.LastmodifiedBy,
+			Attributes:     tempDeveloperDetails.Attributes,
+		}
+	}
+}
+
+func shortListApiProduct(details []ApiProductDetails, verifyApiKeyReq VerifyApiKeyRequest) ApiProductDetails {
+	var bestMathcedProduct ApiProductDetails
+	rankedProducts := make(map[int][]ApiProductDetails)
+	rankedProducts[2] = []ApiProductDetails{}
+	rankedProducts[3] = []ApiProductDetails{}
+
+	for _, apiProd := range details {
+		if len(apiProd.Resources) == 0 || validatePath(apiProd.Resources, verifyApiKeyReq.UriPath) {
+			if len(apiProd.Apiproxies) == 0 || contains(apiProd.Apiproxies, verifyApiKeyReq.ApiProxyName) {
+				if len(apiProd.Environments) == 0 || contains(apiProd.Environments, verifyApiKeyReq.EnvironmentName) {
+					bestMathcedProduct = apiProd
+					return bestMathcedProduct
+					// set rank 1 or just return
+				} else {
+					// set rank to 2
+					rankedProducts[2] = append(rankedProducts[2], apiProd)
+				}
+			} else {
+				// set rank to 3,
+				rankedProducts[3] = append(rankedProducts[3], apiProd)
+			}
+		}
 	}
 
+	if len(rankedProducts[2]) > 0 {
+		return rankedProducts[2][0]
+	} else if len(rankedProducts[3]) > 0 {
+		return rankedProducts[3][0]
+	}
+
+	return bestMathcedProduct
+
+}
+
+func (apiM apiManager) performValidations(dataWrapper VerifyApiKeyRequestResponseDataWrapper) *ErrorResponse {
```

Comment:
As we discussed, it would be better to wrap validation as a method of a specific struct, if there's no outside dependency. This func can perform validation on cross-struct dependencies, and call the validation methods of all substructs (and such recursively).
However, this is just a suggestion, we can leave it as it is now, and improve it in the future.